auth/src/main/java/com/firebase/ui/auth/ui/HelperActivityBase.java

```diff
@@ -84,19 +142,42 @@ public void finish(int resultCode, Intent intent) {
         finish();
     }
 
-    public void saveCredentialsOrFinish(FirebaseUser firebaseUser, IdpResponse response) {
-        saveCredentialsOrFinish(firebaseUser, null, response);
-    }
-
     public void saveCredentialsOrFinish(
             FirebaseUser firebaseUser,
             @Nullable String password,
             IdpResponse response) {
-        SaveSmartLock saveSmartLock = SaveSmartLock.getInstance(this);
-        if (saveSmartLock == null) {
-            finish(Activity.RESULT_OK, response.toIntent());
-        } else {
-            saveSmartLock.saveCredentialsOrFinish(firebaseUser, password, response);
+        mPendingIdpResponse = response;
+
+        String accountType = ProviderUtils.idpResponseToAccountType(response);
+        mSmartLockHandler.saveCredentials(firebaseUser, password, accountType);
+    }
+
+    @RestrictTo(RestrictTo.Scope.TESTS)
+    public SmartLockHandler getSmartLockHandler() {
+        return mSmartLockHandler;
+    }
+
+    private void onPendingResolution(@NonNull PendingResolution resolution) {
+        try {
+            startIntentSenderForResult(
+                    resolution.getPendingIntent().getIntentSender(),
+                    resolution.getRequestCode(),
+                    null, 0, 0, 0);
+        } catch (IntentSender.SendIntentException e) {
+            Log.e(TAG, "STATUS: Failed to send resolution.", e);
+            finish(RESULT_OK, mPendingIdpResponse.toIntent());
```

Comment:
Because this implementation is specific to Smart Lock being failure tolerant, I feel like we should just inline it or add a comment specifying that it will finish the activity should an error occur. What do you think?